java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java

```diff
@@ -260,6 +271,38 @@ public void jsonDeserialize(JSONObject jsonObject) throws PhysicalWebCollectionE
   }
 
   /**
+   * Return a list of UrlGroups sorted by the rank of the top-ranked pair in each group.
+   * @return a sorted list of UrlGroups.
+   */
+  public List<UrlGroup> getUrlGroupsSortedByRank() {
+    List<PwPair> sortedPwPairs = getPwPairsSortedByRank();
+
+    Map<String, UrlGroup> pwGroupMap = new HashMap<>();
+    for (PwPair pwPair : sortedPwPairs) {
+      PwsResult pwsResult = pwPair.getPwsResult();
+      if (pwsResult != null) {
+        String groupId = pwsResult.getGroupId();
+        if (groupId != null && !groupId.equals("")) {
```

Comment:
Correct, we're not including ungrouped results.  It's "getUrlGroupsSortedByRank" so this is by design.  I agree a "get displayable PwPairs" method that sorts and returns the top group results combined with ungrouped results would be more useful.

Currently it's still valid for PWS to return no groupid.  In the future we'll require groupid, correct?  If so, we should validate when we deserialize PwsResult rather than here.  On the other hand if we decide null/empty groupid is valid then we'll still need to filter here.